homeassistant/components/switcher_kis/__init__.py

```diff
@@ -0,0 +1,489 @@
+
+Author: Tomer Figenblat
+
+Example yaml configuration,
+Get you device details using the instructions here:
+https://github.com/NightRang3r/Switcher-V2-Python .
+
+Mandatory only configuration:
+switcher_kis:
+  phone_id: 'xxxx'
+  device_id: 'xxxxxx'
+  device_password: 'xxxxxxxx'
+
+Full configuration:
+switcher_kis:
+  phone_id: 'xxxx'
+  device_id: 'xxxxxx'
+  device_password: 'xxxxxxxx'
+  name: 'my_boiler'
+  friendly_name: "My Friendly Boiler"
+  icon: 'mdi:my-icon'
+  include_schedule_sensors: true
+  schedules_scan_interval:
+    minutes: 5
+
+Defaults:
+name = 'boiler'
+friendly_name = 'Boiler'
+include_sensors = false
+schedules_scan_interval = {'mintues': 5}
+
+The minimal configuration will:
+- Create 1 switch entity called Boiler with the entity id:
+  switch.switcher_kis_boiler.
+- Register the following services:
+  - switcher_kis.turn_on - no arguments needed.
+  - switcher_kis.turn_off - no arguments needed.
+  - switcher_kis.turn_on_15_minutes - no arguments needed.
+  - switcher_kis.turn_on_30_minutes - no arguments needed.
+  - switcher_kis.turn_on_45_minutes - no arguments needed.
+  - switcher_kis.turn_on_60_minutes - no arguments needed.
+  - switcher_kis.set_auto_off - takes 1 string argument {'auto_off': '01:30'}
+  - switcher_kis.update_device_name - takes 1 string argument {'name': 'boil'}
+
+Setting the 'include_schedule_sensors' key to True,
+Will result in all the above plus the following:
+- Creation of 8 sensors entities represnting the device's schedules (0-7),
+  The names will be constructed from the 'name' key '_schedule' and the id,
+  The same goes for the friendly names:
+  - sensor.switcher_kis_boiler_schedule0: Boiler Schedule 0
+  ...
+  - sensor.switcher_kis_boiler_schedule7: Boiler Schedule 7
+
+  The sensors state will allways provide the minumum information:
+  - Not configured
+  - Not enabled
+  - Due today at 17:30
+  - Due tommorow at 20:00
+
+- Register the following services:
+  - switcher_kis.create_schedule - takes the following mandatory arguments:
+    {'start_time': '19:30', 'end_time': '20:30', 'recurring': false}
+    if recurring is set to true, you must provide an extra argument:
+    {'days': ['Monday', 'Wednesday', 'Saturday']}
+  - switcher_kis.delete_schedule - takes 1 int argument {'schedule_id': 3'}
+  - switcher_kis.enable_schedule - takes 1 int argument {'schedule_id': 0'}
+  - switcher_kis.disable_schedule - takes 1 int argument {'schedule_id': 7'}
+
+"""
+
+# pylint: disable=no-name-in-module
+from logging import getLogger
+from typing import Dict, Optional, Any, List  # noqa F401
+from traceback import format_exc
+from asyncio import wait_for, TimeoutError as AsyncioTimeoutError
+from datetime import timedelta, datetime
+
+import voluptuous as vol
+
+from aioswitcher.schedules import SwitcherV2Schedule  # noqa F401
+from aioswitcher.consts import (MONDAY, TUESDAY, WEDNESDAY, THURSDAY,
+                                FRIDAY, SATURDAY, SUNDAY)
```

Comment:
These imports won't work.

The first time the component gets initialized, aioswitcher still needs to get installed, but it won't as loading the component will fail due to a missing import.

As these imports are necessary only for the scheduling sensors part, I think it would be smart to split off the service registration setup into a separate file. (service_registration.py)

Reply:
Thank you again.

I was reluctant to having these imports here, but I had no choice seeing that they are used in the component's configuration.
And I was under the impression that the component won't load until all the REQUIREMENTS are installed.
Perhaps I got it wrong.

I didn't quite get the registration split off solution, can you please elaborate on it?

Reply:
When parsing and loooking for the REQUIREMENTS the component is already loaded.
If you create a separate file with all of the functions from line 249 until 433. And lines 150-200 to a separate file. This file can have as a `DEPENDENCIES=['aioswitcher']` and export a single function `register_services`, which in turn will be called at the end of the `async_setup` function.

This will also help to review the code, and maybe also for testing as you can test the service registration separately.

Reply:
Ok.
I'll look into it.
Thanks.

Reply:
Made the required changes [here](https://github.com/home-assistant/home-assistant/pull/20830/commits/a67dcbfefe43892627db2fc64162a658f5b89d45).
Thank you.